src/components/structures/ScrollPanel.js

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright 2015 OpenMarket Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+var React = require("react");
+var ReactDOM = require("react-dom");
+var GeminiScrollbar = require('react-gemini-scrollbar');
+
+var DEBUG_SCROLL = false;
+
+/* This component implements an intelligent scrolling list.
+ *
+ * It wraps a list of <li> children; when items are added to the start or end
+ * of the list, the scroll position is updated so that the user still sees the
+ * same position in the list.
+ *
+ * It also provides a hook which allows parents to provide more list elements
+ * when we get close to the start or end of the list.
+ *
+ * We don't save the absolute scroll offset, because that would be affected by
+ * window width, zoom level, amount of scrollback, etc. Instead we save an
+ * identifier for the last fully-visible message, and the number of pixels the
+ * window was scrolled below it - which is hopefully be near enough.
+ *
+ * Each child element should have a 'data-scroll-token'. This token is used to
+ * serialise the scroll state, and returned as the 'lastDisplayedScrollToken'
+ * attribute by getScrollState().
+ *
+ * The following properties are exposed:
+ *
+ *   * stickyBottom: if set to true, then once the user hits the bottom of the
+ *     list, any new children added to the list will cause the list to scroll
+ *     down to show the new element, rather than preserving the existing view.
+ *
+ *   * onFillRequest(backwards): a callback which is called on scroll when the
+ *     user nears the start (backwards = true) or end (backwards = false) of the
+ *     list
+ *
+ *   * onScroll: a callback which is called whenever any scroll happens.
+ *
+ *   * className: classnames to add to the top-level div
+ *
+ *   * style: styles to add to the top-level div
+ */
+module.exports = React.createClass({
+    displayName: 'ScrollPanel',
+
+    propTypes: {
+        stickyBottom: React.PropTypes.bool,
+        onFillRequest: React.PropTypes.func,
+        onScroll: React.PropTypes.func,
+    },
+
+    getDefaultProps: function() {
+        return {
+            stickyBottom: true,
+            onFillRequest: function(backwards) {},
+            onScroll: function() {},
+        };
+    },
+
+    componentWillMount: function() {
+        this.savedScrollState = {atBottom: this.props.stickyBottom};
```

Comment:
Nit: I would drop the `saved` from `savedScrollState`. This variable isn't any more "saved" than any other variable. At best, it's a bunch of extra characters. At worst, it's misleading and makes you think it is somehow persisted differently to everything else (e.g. to local storage) which is just not the case. This is simply the scroll state, and you elicit this by exposing the public getter `getScrollState` which just returns this variable.